### ISSUE TITLE
module change subscription callback not called for all events

### ISSUE THREAD

**Issue by adrianlita** [sysrepo/sysrepo#2485]
Hello,

First I want to mention that this happens only in in version 1.4.140.
I call sr_module_change_subscribe() and register a callback, with opts: SR_SUBSCR_CTX_REUSE | SR_SUBSCR_UPDATE. Next, I use a NETCONF client to do some edits on the datastore.

On the first edit that I do through a NETCONF client, everything works as expected (as it does in previous versions), which means that it calls the callback with these events, in order: UPDATE, CHANGE, DONE

On the second edit that I do for the same leaf, it only calls UPDATE, and the rest are not called.

On a third edit, it does not work any more (i assume there is a deadlock on another thread, as my main thread is working).

One other thing to mention is that my callback always returns SR_ERR_OK (for the UPDATE event, for example). And perhaps I'm repeating myself, but in previous version I haven't encountered this error.

Thank you very much!

**Comment by michalvasko** [comment]
Could you post *sysrepo* verbose output? It should be clear from that what exactly is happening.

**Comment by adrianlita** [comment]
[INF]: Processing "nts-manager" "update" event with ID 1 priority 0 (remaining 1 subscribers).
[INF]: Successful processing of "update" event with ID 1 priority 0 (remaining 0 subscribers).
[INF]: Processing "nts-manager" "change" event with ID 1 priority 0 (remaining 1 subscribers).
[INF]: Successful processing of "change" event with ID 1 priority 0 (remaining 0 subscribers).
[INF]: Processing "nts-manager" "done" event with ID 1 priority 0 (remaining 1 subscribers).
[INF]: Successful processing of "done" event with ID 1 priority 0 (remaining 0 subscribers).
[INF]: There are no subscribers for "nts-manager" notifications.
[INF]: Published event "update" "nts-manager" with ID 2 priority 0 for 1 subscribers.
[INF]: Processing "nts-manager" "update" event with ID 2 priority 0 (remaining 1 subscribers).
[INF]: Successful processing of "update" event with ID 2 priority 0 (remaining 0 subscribers).
[INF]: Event "update" with ID 2 priority 0 succeeded.
[INF]: Published event "change" "nts-manager" with ID 2 priority 0 for 1 subscribers.
[INF]: Processing "nts-manager" "change" event with ID 2 priority 0 (remaining 1 subscribers).
[INF]: Successful processing of "change" event with ID 2 priority 0 (remaining 0 subscribers).
[INF]: Event "change" with ID 2 priority 0 succeeded.
[INF]: Published event "done" "nts-manager" with ID 2 priority 0 for 1 subscribers.
[INF]: Processing "nts-manager" "done" event with ID 2 priority 0 (remaining 1 subscribers).
[INF]: Successful processing of "done" event with ID 2 priority 0 (remaining 0 subscribers).
[INF]: There are no subscribers for "nts-manager" notifications.
[INF]: There are no subscribers for "nts-manager" notifications.
[INF]: Processing "nts-manager" "update" event with ID 3 priority 0 (remaining 1 subscribers).
[INF]: Successful processing of "update" event with ID 3 priority 0 (remaining 0 subscribers).


For the first event, it processes UPDATE, CHANGE and DONE.
The second event that you see in the log is an actual edit done from inside the app.

Then the last event is the UPDATE event from when I send another request via NETCONF client.
As we see there, there's no further CHANGE/DONE events.

**Comment by michalvasko** [comment]
Well, I have expected some errors or some output with more information, this all looks fine. So you will probably need to provide the exact use-case, all the steps and data needed to reproduce it, not sure how else can I learn more details.

**Comment by adrianlita** [comment]
Ok, I will try to build something for you to be able to test and get back.

**Comment by adrianlita** [comment]
It is quite strange, as with previous version I don't have any issue with this.
I've just tested, and other module_change subscription I have does not experience this issue.
The only thing I'm doing inside the UPDATE event is changing a leaf back to its original value via sr_set_item().

**Comment by michalvasko** [comment]
> The only thing I'm doing inside the UPDATE event is changing a leaf back to its original value via sr_set_item().

That would explain why no other events are generated since you essentially revert the change. Change and done event are generated only for actual changes.

**Comment by adrianlita** [comment]
> 
> 
> > The only thing I'm doing inside the UPDATE event is changing a leaf back to its original value via sr_set_item().
> 
> That would explain why no other events are generated since you essentially revert the change. Change and done event are generated only for actual changes.

You are, indeed, right; it seems to successfully work when commenting out the sr_set_item() function.
However, if going back just one version, the behavior was totally different.

I don't mind changing my code to accommodate the new behavior, but I just want to make sure that this is the intended behavior.

**Comment by michalvasko** [comment]
> if going back just one version, the behavior was totally different.

I take it the new version is 1.4.140, what was the old version? This is the correct behavior and I do not know about any recent fixes in this regard.

**Comment by adrianlita** [comment]
Yes, new version is 1.4.140 and older version was 1.4.122

**Comment by michalvasko** [comment]
And you are saying that this exact use-case worked differently in 1.4.122? I can only think of the reason being that the edit was not properly detected as the revert of the original one, there could have been some fixes. In any case, the current one is the correct behavor.

**Comment by adrianlita** [comment]
Yes, and not only 1.4.122. I've been having the same exact code for more than 3 months, and constantly updating libyang/sysrepo libs as being released, and only in 1.4.140 the behavior changed. Anyhow, since this is how it's supposed to be, I consider this issue closed. Thank you so much for your help!

**Comment by adrianlita** [comment]
I just have to leave another comment here, because I still feel there's a small bug.

So first of all, if I don't do the sr_set_item(), I don't have issues anymore.

Why I still think it's a bug, is because the behavior is not happening as you mention here. Let me give you my exact scenario:
- consider you have a leaf (uint16) called "started". Current value (not default) is 0.
- via NETCONF, i edit the value to be 1; it is here very important to mention that this is the FIRST request on this leaf
- what happens here is that the callback is called with UPDATE, where I edit the value back to its old value (0)
- then the callback gets called again, with CHANGE and DONE (though you said it shouldn't)
- next, I do another NETCONF edit, and put the value back to 0 (or another value, doesn't matter); now I get the UPDATE callback, where I set the value back to its old value (which was 1).
- then, no other callbacks are received; if I try to get the leaf, it has the value of 1 (which means it was updated via sr_set_item())
- then if I do another NETCONF edit (the third one), absolutely nothing happens (looks like a deadlock or something). so no callbacks, nothing. This is the main reason I still believe there's a bug there.

Sorry for the long message.
